src/components/shared/TreeView.js

```diff
@@ -53,40 +53,138 @@ export type Column<DisplayData: Object> = {|
+    if (ascending === true || current === null) {
+      sortedColumns.push({ column, ascending: true });
+    } else {
+      sortedColumns.push(this._invertSortState(current));
+    }
+    return new ColumnSortState(sortedColumns);
+  }
+
+  _isLastSortedColumn(column: string) {
+    const cur = this.current();
+    return cur !== null && cur.column === column;
+  }
+
+  _invertSortState(state: SingleColumnSortState): SingleColumnSortState {
+    return { column: state.column, ascending: !state.ascending };
+  }
+
+  getStateForColumn(column: string): SingleColumnSortState | null {
+    return this.sortedColumns
+      .filter((c) => c.column === column)
+      .concat(null)[0];
+  }
+
+  getStateForColumnOrDefault(column: string): SingleColumnSortState {
+    return (
+      this.getStateForColumn(column) || { column: column, ascending: true }
+    );
   }
-  return (
-    <div className="treeViewHeader">
-      {fixedColumns.map((col) => (
+
+  current(): SingleColumnSortState | null {
+    return this.sortedColumns.length > 0
+      ? this.sortedColumns[this.sortedColumns.length - 1]
+      : null;
+  }
+
+  sortItemsHelper<T>(items: T[], compareColumn: (T, T, string) => number): T[] {
+    let sorted = items;
+    for (const { column, ascending } of this.sortedColumns) {
+      const sign = ascending ? -1 : 1;
+
+      sorted = sorted.sort((a, b) => {
+        return sign * compareColumn(a, b, column);
+      });
+    }
+    return sorted;
+  }
+}
+
+class TreeViewHeader<DisplayData: Object> extends React.PureComponent<
+  TreeViewHeaderProps<DisplayData>
+> {
+  _onSort = (e: MouseEvent) => {
+    const { onSort } = this.props;
+    const target = e.currentTarget;
+    if (target instanceof HTMLElement) {
+      onSort(Number(target.dataset.column));
+    }
+  };
+
+  render() {
+    const { fixedColumns, mainColumn, currentSortedColumn } = this.props;
+    if (fixedColumns.length === 0 && !mainColumn.titleL10nId) {
+      // If there is nothing to display in the header, do not render it.
+      return null;
+    }
+    return (
+      <div className="treeViewHeader">
+        {fixedColumns.map((col) => {
+          let sortClass = '';
+          if (
+            currentSortedColumn &&
+            currentSortedColumn.column === col.propName
+          ) {
+            if (currentSortedColumn.ascending) {
+              sortClass = 'sortDescending';
+            } else {
+              sortClass = 'sortAscending';
+            }
+          } else {
+            sortClass = 'sortInactive';
+          }
+          return (
+            <PermissiveLocalized
+              id={col.titleL10nId}
+              attrs={{ title: true }}
+              key={col.propName}
+            >
+              <span
+                className={`treeViewHeaderColumn treeViewFixedColumn ${col.propName} ${sortClass}`}
+                data-column={col.propName}
+                onClick={this._onSort}
+              ></span>
```

Comment:
This feature will be extremely useful for screen reader users. Thanks for working on it. However, as they are currently, these sort controls won't be usable by screen reader users:

1. They should have role="button".
2. They should have aria-sort="ascending" if sorted ascending, aria-sort="descending" if sorted descending or no aria-sort attribute if not sorting by this column.
3. They should have tabindex="0" to make them focusable. Ideally, the sort controls would all occupy a single tab stop and you'd use the left and right arrows to move between them, but that will take more work. In the interim, I think it's reasonable to have them all reachable with tab. If these extra tab stops are a real problem for non-keyboard screen reader users, although not ideal, this could be dealt with in a follow-up; screen reader users can get to these controls without them being focusable.